### PR TITLE
VM side of synthetic handler fix

### PR DESCRIPTION
**Pull request by gacholio**
Details in #9267

Also reverts #7340 which was a failed attempt to fix this before we
completely understood the issue.

[ci skip]

Signed-off-by: Graham Chapman <graham_chapman@ca.ibm.com>

### CODE REVIEW

**Comment by gacholio** [comment]
Note that this PR by itself is slightly incorrect - if we encounter an exception catch whose classred is at the newly-invalid CP indices, things will go wrong.

This is so unlikely to be noticed in the next couple of days that I don't think it's worth staging this properly.

**Comment by gacholio** [comment]
@cathyzhyi So you can see when this has been merged.

**Comment by gacholio** [comment]
jenkins test sanity all jdk8

**Comment by gacholio** [comment]
The changes in `c_jitDecompileAtExceptionCatch` are entirely the revert of the previous attempt.

**Comment by gacholio** [comment]
I see a mistake in the 16-bit path in MethodMetadata.c - not using the local catchType. I'll fix this post-review to avoid invalidating all of the testing that happened.

**Comment by gacholio** [comment]
As this isn't being reviewed tonight, I've made the fix.

**Comment by gacholio** [comment]
jenkins test sanity,extended all jdk8

**Comment by gacholio** [comment]
jenkins test sanity,extended all jdk8

**Comment by gacholio** [comment]
AIX failure is infra-related (the previous one passed).

File: runtime/codert_vm/decomp.cpp [hunk 1]
```diff
@@ -1200,12 +1200,45 @@ c_jitDecompileAtExceptionCatch(J9VMThread * currentThread)
...
+	UDATA newNumberOfFrames = 1;
+	void *stackMap = NULL;
+	void *inlineMap = NULL;
+	void *inlinedCallSite = NULL;
+	/* Note we need to add 1 to the JIT PC here in order to get the correct map at the exception handler
+	 * because jitGetMapsFromPC is expecting a return address, so it subtracts 1.  The value stored in the
+	 * decomp record is the start address of the compiled exception handler.
+	 */
+	jitGetMapsFromPC(vm, metaData, (UDATA)jitPC + 1, &stackMap, &inlineMap);
+	Assert_CodertVM_false(NULL == inlineMap);
```
Comment (DanHeidinga) [inline runtime/codert_vm/decomp.cpp:1232]: Are we guaranteed to have an inlineMap even if nothing was inlined here?  Should this assert be moved inside the if?

Comment (gacholio) [reply runtime/codert_vm/decomp.cpp:1232]: This assert has never fired, in testing or production, so I'll say it's correct.

File: runtime/codert_vm/decomp.cpp [hunk 2]
```diff
@@ -1200,12 +1200,45 @@ c_jitDecompileAtExceptionCatch(J9VMThread * currentThread)
 	U_8 *jitPC = decompRecord->pc;
 	/* Simulate a call to a resolve helper to make the stack walkable */
 	buildBranchJITResolveFrame(currentThread, jitPC, J9_STACK_FLAGS_JIT_EXCEPTION_CATCH_RESOLVE);
-	/* Discard the existing decompilation in favour of a new one at the exception catch point */
+
+	J9JavaVM *vm = currentThread->javaVM;
+	J9JITDecompileState decompileState;
```
[the remark this answers is not in the document]
Comment (gacholio) [reply runtime/codert_vm/decomp.cpp:1205]: This (and the code below) are both from the revert of the previous attempt. Not sure it's worth changing anything.

[the remark this answers is not in the document]
Comment (gacholio) [reply runtime/codert_vm/decomp.cpp:1205]: This certainly isn't a performance critical path, so if you really want the zeroing, I'll do it (I'm just never sure I trust the compiler to optimize this when most/all of the fields get stored anyway).

[the remark this answers is not in the document]
Comment (DanHeidinga) [reply runtime/codert_vm/decomp.cpp:1205]: Gac pointed out that this is just code being moved, not new code.

File: runtime/codert_vm/decomp.cpp (hunk 1 above)
Comment (DanHeidinga) [reply runtime/codert_vm/decomp.cpp:1232]: And GAC pointed out above (and I missed):
> The changes in c_jitDecompileAtExceptionCatch are entirely the revert of the previous attempt.

**Pull request merged**